### PR TITLE
Add ability to perform an etcd on-demand snapshot via cli

### PR DESCRIPTION
**Pull request by briandowns**
Signed-off-by: Brian Downs <brian.downs@gmail.com>

#### Proposed Changes ####

This PR adds a new sub-command to the K3s command called `etcd-snapshot` and it takes a number of the standard K3s flags as well as specific flags like `snapshot-name`, used to give the snapshot a specific name and `dir` used to tell K3s where to save the snapshot. The former use applicable in this use case as well as for cron scheduled snapshots and keeps in like with RKE features for snapshotting and will ultimately be pulled into RKE2. An on-demand snapshot will effect snaptshot retention (if saved to the snapshot directory) and will result in a check to see if an older snapshot needs to be purged.

#### Verification ####

1. Start K3s using Etcd and wait for it to get to steady state. 
2. Run `k3s etcd-snapshot` to perform an etcd snapshot.
  A. Verify the snapshot has been performed successfully via log output `{"level":"info","msg":"saved","path":"/var/lib/rancher/k3s/server/db/snapshots/etcd-snapshot-<unix-timestamp>"}`
  B. Verify that the newly created snapshot is indeed at the expected location
3. Run `k3s etcd-snapshot --snapshot-name=my_snapshot` to perform and etcd snapshot now.
  A. Perform the same verification steps from above.

#### Linked Issues ####

#2758

### CODE REVIEW

File: pkg/cluster/cluster.go
Comment (brandond) [inline pkg/cluster/cluster.go:42]: I think Go style guide usually drops a trailing else:
```suggestion
		}
		os.Exit(0)
```

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:743]: 💯

**Review by brandond — DISMISSED** [review]

**Comment by brandond** [comment]
Based steps 1 and 2 of your verification steps, it sounds like K3s needs to be running when you take the on-demand snapshot. However, a review of the code suggests that it should NOT be running since the snapshot occurs partway through the normal startup process and would conflict with a running K3s server process?

Do you have a corresponding docs PR ready to go along with the feature?

**Comment by briandowns** [comment]
> Based steps 1 and 2 of your verification steps, it sounds like K3s needs to be running when you take the on-demand snapshot. However, a review of the code suggests that it should NOT be running since the snapshot occurs partway through the normal startup process and would conflict with an already-running K3s server process?
> 
> Do you have a corresponding docs PR ready to go along with the feature?

K3s needs to be running for backup to occur since it depends on an active etcd initializedc (client, etc) system. The code is in a place where it wouldn't conflict with an already running system since it's past initialization but before listeners are setup.

File: pkg/cluster/cluster.go
Comment (briandowns) [reply pkg/cluster/cluster.go:42]: Yeah, it does. Remnants of previous edits. Nice catch.

**Comment by brandond** [comment]
> K3s needs to be running for backup to occur since it depends on an active etcd initializedc (client, etc) system. The code is in a place where it wouldn't conflict with an already running system since it's past initialization but before listeners are setup.

OK, nice! It's a bit odd since this is the only time you'd ever want to run two `k3s server` or `k3s agent` processes at the same time, but we've gone around enough times on the new command discussion so it's not worth getting into again.

What's the UX if K3s isn't running when you run the command?

**Comment by briandowns** [comment]
> What's the UX if K3s isn't running when you run the command?

Dial context cancellation fires and errors that it cant connect.

**Comment by cjellick** [comment]
From the requirements:

> Exposing on-demand snapshots through an integration with Rancher2. This means on-demand snapshots WILL NOT be exposed in the rke2 cli. You will only get on-demand snaps through Rancher2 integration.

I havent looked at the code, but based on the description, it seems that this PR _will_ expose it as cli options. Is that right?

Lets talk about it in today's design review.

**Review by cjellick — DISMISSED** [review]
See my above comment, let's hold off on merging until we discuss the approach of adding it to the CLI.

File: scripts/package-cli
Comment (nickgerace) [inline scripts/package-cli:10]: Do you need `k3s` here as well?

Comment (briandowns) [reply scripts/package-cli:10]: No, it's calling back to the server binary.

File: cmd/k3s/main.go
Comment (brandond) [inline cmd/k3s/main.go:38]:
```suggestion
		cmds.NewEtcdSnapshotCommand(wrap(version.Program+"-etcdsnapshot", dataDir, os.Args)),
```

File: scripts/package-cli
Comment (brandond) [inline scripts/package-cli:10]:
```suggestion
for i in crictl kubectl k3s-agent k3s-server k3s-etcdsnapshot; do
```

File: pkg/cli/cmds/etcd_snapshot.go
Comment (brandond) [inline pkg/cli/cmds/etcd_snapshot.go:8]: We're building this to be used as `k3s etcdsnapshot` right? The binaries are called `k3s-etcdsnapshot` so shouldn't this be:
```suggestion
const EtcdSnapshotCommand = "etcdsnapshot"
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: cmd/k3s/main.go
Comment (brandond) [inline cmd/k3s/main.go:38]: It's ugly, but server and agent don't have a const for the command string so you could drop it if you want, or do it like this:
```suggestion
		cmds.NewEtcdSnapshotCommand(wrap(version.Program+"-"+cmds.EtcdSnapshotCommand, dataDir, os.Args)),
```

File: pkg/cli/cmds/etcd_snapshot.go
Comment (brandond) [inline pkg/cli/cmds/etcd_snapshot.go:13]:
```suggestion
		Usage:           "Trigger an immediate etcd snapshot",
```

File: pkg/cli/cmds/etcd_snapshot.go
Comment (brandond) [inline pkg/cli/cmds/etcd_snapshot.go:27]: I like the idea you had here - `k3s server --etcd-snapshot-name=x` becomes `k3s etcd-snapshot --name=x` but changing the flag names means that it breaks setting these things via the config yaml. They need to be the same across all commands.

For example, I might have `etcd-snapshot-dir: /opt/etcd-snapshots/` in /etc/k3s/config.yaml, and would expect `k3s etcd-snapshot` to honor that setting, but it will not because the flag on this command is just `dir`.

While you're at it, I think we should probably also expose the --etcd-snapshot-retention flag here.

**Review by brandond — CHANGES_REQUESTED** [review]
The failure UX for this could use a little work:

* If I run it on a clean install, or on an agent, I get:
    ```
    INFO[2021-01-12T14:30:37.356484384-08:00] Managed etcd cluster initializing
    ERRO[2021-01-12T14:30:37.356558383-08:00] open /var/lib/rancher/k3s/server/tls/etcd/client.crt: no such file or directory
    ```
* If I run it on a system that is using kine, it claims it is initializing etcd, and then hangs until I hit control-c:
    ```
    INFO[2021-01-12T14:42:07.004508051-08:00] Managed etcd cluster initializing
    ^C
    {"level":"warn","ts":"2021-01-12T14:43:10.275-0800","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://127.0.0.1:2379","attempt":0,"error":"rpc error: code = Canceled desc = latest balancer error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
ERRO[2021-01-12T14:43:10.275311725-08:00] failed to check etcd status for snapshot: context canceled
    ````
* If I run it on a system that has been bootstrapped into an etcd cluster but `k3s server` isn't running, it just silently hangs forever? The snapshot command should probably be run in a deadlined context, or with a new etcd client that has the reconnection logic disabled or tuned down.

File: go.mod
Comment (brandond) [inline go.mod:85]: Curious what this got pulled in by

Comment (briandowns) [reply go.mod]: Removed.

**Review by brandond — APPROVED** [review]
LGTM once CI is clean

**Pull request closed without merging**
